src/valid/handles.rs

```diff
@@ -0,0 +1,631 @@
+            }
+            crate::Expression::ImageQuery { image, query } => {
+                handle.check_dep(image)?;
+                match query {
+                    crate::ImageQuery::Size { level } => {
+                        handle.check_dep_opt(level)?;
+                    }
+                    crate::ImageQuery::NumLevels
+                    | crate::ImageQuery::NumLayers
+                    | crate::ImageQuery::NumSamples => (),
+                };
+            }
+            crate::Expression::Unary {
+                op: _,
+                expr: operand,
+            } => {
+                handle.check_dep(operand)?;
+            }
+            crate::Expression::Binary { op: _, left, right } => {
+                handle.check_dep(left)?.check_dep(right)?;
+            }
+            crate::Expression::Select {
+                condition,
+                accept,
+                reject,
+            } => {
+                handle
+                    .check_dep(condition)?
+                    .check_dep(accept)?
+                    .check_dep(reject)?;
+            }
+            crate::Expression::Derivative {
+                axis: _,
+                expr: argument,
+            } => {
+                handle.check_dep(argument)?;
+            }
+            crate::Expression::Relational { fun: _, argument } => {
+                handle.check_dep(argument)?;
+            }
+            crate::Expression::Math {
+                fun: _,
+                arg,
+                arg1,
+                arg2,
+                arg3,
+            } => {
+                handle
+                    .check_dep(arg)?
+                    .check_dep_opt(arg1)?
+                    .check_dep_opt(arg2)?
+                    .check_dep_opt(arg3)?;
+            }
+            crate::Expression::As {
+                expr: input,
+                kind: _,
+                convert: _,
+            } => {
+                handle.check_dep(input)?;
+            }
+            crate::Expression::CallResult(function) => {
+                Self::validate_function_handle(function, functions)?;
+            }
+            crate::Expression::AtomicResult { .. } => (),
+            crate::Expression::ArrayLength(array) => {
+                handle.check_dep(array)?;
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_block_handles(
+        block: &crate::Block,
+        expressions: &Arena<crate::Expression>,
+        functions: &Arena<crate::Function>,
+    ) -> Result<(), InvalidHandleError> {
+        let validate_block = |block| Self::validate_block_handles(block, expressions, functions);
+        let validate_expr = |handle| Self::validate_expression_handle(handle, expressions);
+        let validate_expr_opt = |handle_opt| {
+            if let Some(handle) = handle_opt {
+                validate_expr(handle)?;
+            }
+            Ok(())
+        };
+
+        block.iter().try_for_each(|stmt| match stmt {
```

Comment:
Similarly: please `match *stmt` and simplify all the patterns.

Reply:
Changed as suggested with 62f43a66.